riscv/mmu.cc

```diff
@@ -161,6 +161,143 @@ void mmu_t::load_slow_path(reg_t addr, reg_t len, uint8_t* bytes, uint32_t xlate
   }
 }
 
+void mmu_t::clean_inval(reg_t addr, bool clean, bool inval)
+{
+  reg_t paddr = addr;
+  bool page = false, guest = false, access = false;
+  bool has_gva;
+  reg_t gpa;
+  access_type type = LOAD;
+  bool fault = true;
+  try {
+    paddr = translate(addr, 1, LOAD, 0);
+    type = LOAD;
+    fault = false;
+  } catch (trap_load_access_fault& t) {
+    access = true;
+    has_gva = t.has_gva();
+  } catch (trap_load_page_fault& t) {
+    page = true;
+    has_gva = t.has_gva();
+  } catch (trap_load_guest_page_fault& t) {
+    guest = true;
+    gpa = t.get_tval2();
+  }
+
+  if (fault) {
+    try {
+      paddr = translate(addr, 1, STORE, 0);
+      type = STORE;
+      fault = false;
+    } catch (trap_store_access_fault& t) {
+      access = true;
+      has_gva = t.has_gva();
+    } catch (trap_store_page_fault& t) {
+      page = true;
+      has_gva = t.has_gva();
+    } catch (trap_store_guest_page_fault& t) {
+      guest = true;
+      gpa = t.get_tval2();
+    }
+  }
+
+  if (fault) {
+    try {
+      paddr = translate(addr, 1, FETCH, 0);
+      type = FETCH;
+    } catch (trap_instruction_access_fault& t) {
+      access = true;
+      has_gva = t.has_gva();
+    } catch (trap_instruction_page_fault& t) {
+      page = true;
+      has_gva = t.has_gva();
+    } catch (trap_instruction_guest_page_fault& t) {
+      guest = true;
+      gpa = t.get_tval2();
+    }
+  }
+
+  if (access) {
+    trap_store_access_fault(has_gva, addr, 0, 0);
+  } else if (guest) {
+    trap_store_guest_page_fault(addr, gpa, 0);
+  } else if (page) {
+    trap_store_page_fault(has_gva, addr, 0, 0);
+  } else {
+    if (auto host_addr = sim->addr_to_mem(paddr)) {
+      if (tracer.interested_in_range(paddr, paddr + PGSIZE, type))
+        tracer.clean_invalidate(paddr, blocksz, clean, inval);
+      else
+        refill_tlb(addr, paddr, host_addr, type);
+    } else {
+      throw trap_store_access_fault((proc) ? proc->state.v : false, addr, 0, 0);
+    }
+
+    if (!matched_trigger) {
+      matched_trigger = trigger_exception(type == LOAD? OPERATION_LOAD :
+                                          type == STORE? OPERATION_STORE : OPERATION_EXECUTE,
+                                          addr, 0);
+      if (matched_trigger)
+        throw *matched_trigger;
+    }
+  }
+}
+
+  void mmu_t::cbo_zero(reg_t addr) {
+    addr &= ~(blocksz - 1);
+    reg_t vpn = addr >> PGSHIFT;
+    if (likely(tlb_store_tag[vpn % TLB_ENTRIES] == vpn)) {
```

Comment:
There is a lot of logic here (and in clean_inval() too) to perform what is basically a simple store, no? There ought to be a better way than to have to deal with all these possibilities (TLB hit/miss, tracing, triggers) over and over.

Reply:
Yes, this is similar to simple store, However the store length is longer. So the effect of TLB hit/miss, tracing, triggers is a little different from simple store.

Reply:
Sorry, I cannot get your idea of a better way. Current MACRO for load/store cannot fit into block size operation.

Reply:
@aswaterman any ideas? cbo_zero() and clean_inval() ought to be much simpler for what they do.

Reply:
@scottj97 I agree.

`cbo.zero` was defined in such a way that breaking it into a sequence of smaller stores is legal.  IMO, we should take advantage of that property here.  Could delete this routine and replace the instruction implementation with something like

    auto base = RS1 & -blocksz;
    for (size_t offset = 0; offset < blocksz; offset += 4)
      MMU.store_uint32(base + offset, 0);

Reply:
Better yet, increment "offset" by 1 and simply store bytes, since that's the arch definition (and avoids assumptions about the atomicity of the individual operations).

Reply:
Yeah, that would get the point across better.  They should be equivalent as far as Spike's concerned, though, since the difference in atomicity isn't observable--Spike's single-threaded, and 4 bytes is the smallest protection granule.

Reply:
OK. I still have one problem: How about the tracer for cache_misses and write_accesses counter? Do we need to take it into acount?

Reply:
I think this approach is OK with respect to those counters.  They're part of the Spike implementation, rather than governed by the ISA, and so we can just say that the current behavior is correct by definition.

Reply:
OK.